### DISCUSSION TITLE
Is it possible specify "additionalProperties: false" on OpenAPI types?

### DISCUSSION THREAD

**Discussion by jamestelfer** [goadesign/goa#2833] [Q&A]
I can't see a way to do this currently either with the DSL or via metadata.

Is there an easy way to do this, or is there an example of a plugin that modifies the OpenAPI documents? (TBH, a pointer at the hook to plug into would be just as handy).

Thanks!

**Comment by raphael** [comment]
Hmm I thought that omitting this field would be equivalent to setting it to false. If that's not the case we could remove the [omitempty tag on the struct field](https://github.com/goadesign/goa/blob/3efab9fff655bee28b396a8502d2edf3d128d264/http/codegen/openapi/json_schema.go#L48) so it always generates even when false.

**Reply by jamestelfer** [reply]
Noodling around the specs ... Goa supports OpenAPI 3.03, whose [specification states](https://github.com/OAI/OpenAPI-Specification/blob/main/versions/3.0.3.md#properties) that:

> additionalProperties - Value can be boolean or object. Inline or referenced schema MUST be of a Schema Object and not a standard JSON Schema. Consistent with JSON Schema, additionalProperties defaults to true.

So that would mean changing the schema definition as you suggest I think.

It leads to another question though: if the API spec suggests that additional properties are not allowed, should the server code reject requests that include additional properties instead of ignoring them? I'd say probably not (or at least opt in) as it would make it harder for services to evolve their APIs by adding additional capabilities.

I'm OK with a PR for modifying the spec generation, but I wouldn't really know where to start with changing the code generation for the second part.

**Reply by raphael** [reply]
I see yes we should remove the `omitempty` tag then - thanks for pointing this out. I guess I don't see how one would specify additional properties in the first place unless the type was a `MapOf` in which the value of `additionalProperties` should be true?

**Reply by jamestelfer** [reply]
If a request is created by a client in a different language, it could send a bunch of unsupported properties. Should the server actively reject such a request?

It's not valid according to the schema, so a strict interpretation would be "no it shouldn't". Practically though, it's often an advantage to accept such requests and ignore the additional data.

I'm happiest with leaving it as it is, just flagging it as a scenario to consider.

**Reply by jamestelfer** [reply]
I'll create an issue and a PR for the change above, thanks for the direction!

**Reply by raphael** [reply]
Yep, makes sense. Thank you for putting together a PR!
